### ISSUE TITLE
[RFC]: Extra sampling params support in REST API

### ISSUE THREAD

**Issue by afeldman-nm** [vllm-project/vllm#17191]
### Motivation.

Improve OpenAI API compliance by requiring “extra” sampling params (those sampling params which are not part of the OpenAI specification) to be specified as key-value pairs within an `extra_sampling_params` dict argument.

This is ultimately a prerequisite for logits processor support, since logits processors will be configured through arguments passed to `extra_sampling_params`.

### Proposed Change.

If you are using the HTTP client, then any sampling parameters which are not part of the OpenAI specification for the particular endpoint (“extra sampling params”) will only be possible to set via the `extra_sampling_params` argument

```
extra_sampling_params: Optional[dict[str, Any]]
```

in which the sampling param in question will be passed in as a key-value pair. So for example, the usage of `ignore_eos` in the following HTTP client request

```
curl http://0.0.0.0:8000/v1/completions \
-H "Content-Type: application/json" \
-d '{"model": "facebook/opt-125m", "prompt": "Say this is a test", “ignore_eos": true}'
```

will become

```
curl http://0.0.0.0:8000/v1/completions \
-H "Content-Type: application/json" \
-d '{"model": "facebook/opt-125m", "prompt": "Say this is a test", “extra_sampling_params”: {“ignore_eos": true}}’
```

The end-goal is that setting extra sampling params directly (i.e. not as key/value pairs in extra_sampling_params) will no longer be possible. However, this will break existing code which does not use `extra_sampling_params` to set fields such as `ignore_eos` which are not part of the OpenAI API. So I propose that in the near term, we should allow two ways to set these extra sampling params: either directly as REST API arguments (currently approach), or else as key/value pairs in the `extra_sampling_params` REST API argument. We should wait for a “major” release (in whatever appropriate sense of the word “major”) before forcing users to rely on `extra_sampling_params`.

This RFC only applies to API endpoints which trigger sampling, summarized below along with their associated request classes in `protocols.py` and a link to the associated OpenAI API documentation which defines supported sampling params:

/v1/completions (`CompletionRequest`)
API documentation: https://platform.openai.com/docs/api-reference/completions/create 
/v1/chat/completions (`ChatCompletionRequest`)
API documentation: https://platform.openai.com/docs/api-reference/chat 
/v1/audio/transcriptions (`TranscriptionRequest`)
API documentation: https://platform.openai.com/docs/api-reference/audio/createTranscription 

The following API endpoints do not trigger sampling and are not part of this workstream (note that to save time in writing this RFC, I refer to the endpoints in terms of broad categories here):
Embeddings (`EmbeddingCompletionRequest`, `EmbeddingChatRequest`)
Rerank (`RerankRequest`)
Tokenization/Detokenization (`TokenizationCompletionRequest`, `TokenizationChatRequest`, `DetokenizeRequest`)
LoRA load (`LoadLoRAAdapterRequest`) and unload (`UnloadLoRAAdapterRequest`)

If you are using the OpenAI Python SDK (or similar SDK in another language), the completion/chat/transcription request method does not have an `extra_sampling_params` argument; `extra_sampling_params` will need to be passed in as a key-value pair to the `extra_body` dict argument of the request method. Note that the `extra_body` argument is not part of the server’s REST API and if you pass `extra_body` as an argument within an HTTP client request, the server will ignore it. `extra_body` is simply a “catch-all” argument supported by the Python SDK to handle “special” parameters. Internally, the SDK unpacks `extra_body` into REST API arguments. The server does not see the `extra_body` argument.

So for example, the following SDK request

```
   completion = await client.completions.create(model=model_name,
                                                prompt="Hello, my name is",
                                                max_tokens=5,
                                                temperature=0.0,
                                                extra_body={“ignore_eos”: True)
```

Becomes

```
   completion = await client.completions.create(model=model_name,
                                                prompt="Hello, my name is",
                                                max_tokens=5,
                                                temperature=0.0,
                                                extra_body={“extra_sampling_params”:           
                                                            {“ignore_eos”: True})
```

Plan for rolling out extra sampling params:

### PR 1 ( #16862 , although this PR is WIP is does not yet satisfy the specifications below) 
* Define a list of “extra” sampling params for all three relevant API endpoints
* For each relevant API endpoint, add an `extra_sampling_params member` which may only be used to configure the "extra" sampling params associated with that endpoint
* As of PR 1, it will still be possible to configure these extra sampling params directly as REST API arguments. 
* If a param is configured redundantly or in a conflicting fashion (i.e. if it is specified directly as a REST API argument and also via `extra_sampling_params`) the value in `extra_sampling_params` will have precedence.
* This PR is a prerequisite for near-term work on logits processor support
* This PR does not introduce breaking changes

### PR 2 (recommend aligning with major release)
* It is no longer possible to configure extra sampling params directly as REST API arguments (breaking change)
* This PR brings completeness/correctness for the extra sampling params workstream, since there should only be one way to set extra sampling params

### Feedback Period.

1 week

### CC List.

@njhill @comaniac @WoosukKwon @simon-mo 

CC @robertgshaw2-redhat 

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by russellb** [comment]
An opaque passthrough for use by custom logits processors makes sense to me.

One thing we need to be careful of is (at least IMO) we should NOT use an opaque field like this for parameters that are supported by code included in vLLM. I think we are better served by declaring stronger (and clearer) types for that. Perhaps during implementation, you could leave some comments along these lines to encourage others in the future not to abuse the field.

**Comment by russellb** [comment]
Something I think we should consider here and for any future items we're adding to the body of an OpenAI request: We should namespace them somehow to avoid possible conflicts with official parameters added by OpenAI in the future. For example, instead of `extra_sampling_params`, do something with a `vllm` prefix, like `vllm_sampling_params`.

**Comment by njhill** [comment]
Thanks @afeldman-nm I think this is good but am still a bit unsure. Having a nested `extra_sampling_params` in the request json is pretty verbose/cumbersome, e.g. for simple curl examples etc. We could mitigate that by:
- Using a shorter name e.g. just `"extras"`. TBH I think "sampling_params" in the name is kind of superfluous anyhow.
- Instead of nested, have custom args be top-level but with a rule that their names must begin with e.g. `x_`, so that they won't conflict with any future built-in args.

For the same reason (as well as for backwards-compatibility of course), I don't think we should move the built-in vllm-specific params. Though I also understand @russellb's point about this.

**Comment by russellb** [comment]
I'm not suggesting moving any existing params, but I think namespacing our custom params is the right thing to do. We don't own that namespace, openai does. They could step on anything we're using at any time. Namespacing would avoid that.

**Comment by russellb** [comment]
> Having a nested extra_sampling_params in the request json is pretty verbose/cumbersome

but it should be rare, right? It would only be relevant to custom out-of-tree code (or it should be)

**Comment by russellb** [comment]
some thoughts

* This is specific to the custom logits processors plugin interface, right? If so, perhaps a more desscriptive name would be appropriate. Instead of `vllm_extras`, `vllm_logits_extras` or something -- I don't have strong opinions on the exact name, just that it's clear what it applies to and resides in the vllm namespace.

* Would multiple logits processor plugins be supported at the same time? If so, how is one (or more) selected? How would you deal with the different plugins having different schemas for its inputs?

Another way to deal with the validation side would be to add a validation callback as part of the plugin interface. Let the plugins deal with it and developers/users of multiple plugins can also deal with conflicts between the validations. I don't think core vllm needs to work overly hard at that.

**Comment by afeldman-nm** [comment]
> Based on feedback I propose the following RFC revision:
> 
> * Inside `CompletionRequest`, `ChatCompletionRequest`, and `TranscriptionRequest`, `extra_sampling_params` becomes `vllm_extras` for improved brevity
> * Rather than making `vllm_extras` completely opaque (as is currently the case in my RFC proposal & as [@russellb](https://github.com/russellb) suggested), we can implement `vllm_extras` in a fashion that will be forward-compatible with a plugin-based logits processor interface: (1) By the time vLLM initialization is finished, the engine contain a pydantic model which embodies all valid `vllm_extras` arguments, and during initial request processing `vllm_extras` should be validated against the Pydantic model. (2) Eventually (once logits processor support is fully-realized) the Pydantic model should be assembled dynamically at vLLM startup, by aggregating required arguments from all logits processor plugins. (3) This implies that each logits processor plugin includes a dict of required `vllm_extras` arguments and their type signatures.
> * However, do not actually implement plugin support as part of this "custom args" workstream; implement the skeleton of custom args support as part of this workstream, and then leave the plugin implementation to the "logits processor interface" workstream (RFC: [[RFC][V1] `LogitsProcessor` interface #13360](https://github.com/vllm-project/vllm/pull/13360) ). More detail below.
> 
> Regarding the process for logits processor plugins to extend the pydantic model that validates `vllm_extras`, here is an example of how a Pydantic model can be extended with additional arguments:
> 
> ```
> PluginField = Tuple[type, Any]
> 
> class UserModel(BaseModel):
>     id: int
>     name: str
>     email: str
> 
> plugin_fields: Dict[str, PluginField] = {
>     "age": (int, Field(..., ge=0, description="User age must be ≥ 0")),
>     "favourite_colour": (str, Field(..., min_length=1)),
> }
> 
> create_model(
>         f"{core.__name__}WithPlugins",
>         __base__=UserModel,
>         **plugin_fields, 
>     )
> ```
> 
> I like this approach because `vllm_extras` arguments are validated but also extensible which gives us an easy way to prevent core vllm args from being passed in.
> 
> That being said, we are not quite yet at the point of having the logits processor plugin interface fully defined and implemented (separate PR), and I believe that all of the "core" logits processors which [@njhill](https://github.com/njhill) defined in [#13360](https://github.com/vllm-project/vllm/pull/13360) rely on core vLLM arguments such as `min_p` which are not part of this "custom sampling params" workstream. So for the purposes of landing custom sampling params soon & decoupling from the logits processor interface workstream, I'll probably add a fixed `CustomArgsModel` pydantic model with no members to the core vLLM code and vLLM will validate `vllm_extras` against this model. Of course this means that in the near term vLLM API will only accept an empty `vllm_extras` dict. However, for the purposes of unit-testing the custom args feature, it should be possible to use `monkeypatch.setattr()` or something similar to inject fake custom args into `CustomArgsModel`.
> 
> Once the logits processor interface is fully fleshed out, then it can be part of that workstream to replace the empty `CustomArgsModel` definition with a dynamically-constructed model derived from plugins.
> 
> Would love to know what you think of this and if this leaves any concerns unaddressed?
> 
> CC: [@njhill](https://github.com/njhill) [@russellb](https://github.com/russellb) [@simon-mo](https://github.com/simon-mo) but also open to feedback from anyone else

Alright, closing comment on this RFC since it has been greater than a week. Thanks everyone for the feedback.

I plan to move forward with the approach described in my previous message, with the following changes:
* In the REST API interface, the custom args argument will be `vllm_xargs`
* The type signature for this custom args argument will be `dict[str,CustomArgValueType]` where `CustomArgValueType = Union[int,str]`. This is a temporary solution which ensures that it is easy for pydantic to validate the API argument. However, it is likely that subsequent PRs will modify this interface to support a wider range of custom arguments

**Comment by afeldman-nm** [comment]
Resolved by #16862
